### ISSUE TITLE
Should I commit the dependencies in my vendor directory?

### ISSUE THREAD

**Issue by chx** [composer/composer#10923]
https://getcomposer.org/doc/faqs/should-i-commit-the-dependencies-in-my-vendor-directory.md is misleading and as shown in the recent plugin catastrophe leads to broken builds.

Let's go step by step:

> The best practice is to then have all the developers use Composer to install the dependencies.

This has multiple problems: a) everyone on the team now needs to have and use Composer. That's very often unnecessary if this guidance is not followed b) it's slower than git. This is not to diss on Composer, it has made great improvement in speed but beating git is beyond what any PHP script can possibly do. 

> Similarly, the build server, CI, deployment tools etc should be adapted to run Composer as part of their project bootstrapping.

This can break the build when a) one of the repositories being pulled from is unresponsive b) one of the repositories being pulled from has a malicious attack c) composer itself breaks (let's hope this one doesn't happen again any time soon).

> Large VCS repository size and diffs when you update code.

Disk space is, by and large, free. Although this was not Jim Gray's seminal 2006 (!) presentation title, the rule of thumb for long has been "Disk is the new tape". Most of composer managed files are _text_ anyways. git has added pathspec magic `:(exclude)` and its short form `:!` in 2014, so excluding `vendor` from diff is trivial. And you actually might need to diff vendor code -- what if your bug is actually in vendor or it's an interaction with vendor code? It's not magic, it's just software. It has bugs. (Sometimes, it's my code. Of course it has bugs.)

> Duplication of the history of all your dependencies in your own VCS.

This is again at best a half truth: only the tags you are actually using will be duplicated. Not every commit, not every branch.

> Adding dependencies installed via git to a git repo will show them as submodules. This is problematic because they are not real submodules, and you will run into issues.

Yeah, just remove `.git` before adding them. Next time composer runs -- which will be infrequent anyways after this -- it will just clone from git anew. To quote my own `composer.json`:

    "scripts": {
        "remove-git-submodules": "find vendor -type d -name .git | xargs rm -rf",

**Comment by Seldaek** [comment]
> it's slower than git

Yes and no, once your repo is 5GB worth of vendor dir diffs, cloning it isn't going to be fast.

Try a git clone https://github.com/phpstan/phpstan to see what I mean.

> Next time composer runs -- which will be infrequent anyways after this

I think you should probably run composer updates more often than you seem to be doing.

----

The rest is mostly opinions, you disagree with our recommendations, feel free to do it your way.

**Comment by stof** [comment]
> 2\. what if the bug you are currently working on is actually in vendor or it's an interaction with vendor code?

In that case, you should probably not edit the code in `vendor` to fix the bug (otherwise, you'll loose the bugfix the next time you run composer).
And if you fix the bug by updating to a newer version with Composer, the composer.lock file will also be modified (and this one will appear in the git diff)

**Comment by stof** [comment]
`composer require` locks all existing packages by default. And `composer update` allows to update a few packages if you want instead of doing a full update. So those "minimal run" are still possible.

But all this can still work by committing only the lock file rather than the whole vendor directory.

**Comment by chx** [comment]
I use `composer require`, of course, in fact, I never use `composer update` itself exactly because of this, however that only works in the simplest of case: when there are no dependencies needed to be updated. The moment you do need to update, you do not really have any options afaik because `composer require -W` will do a maximal and there's no "install this package, fix dependencies if needed and stop" option.

**Comment by ksenzee** [comment]
I have always felt the documentation is doing newcomers a disservice by keeping it a secret that many people do commit their vendor directory by choice. Shouldn't the documentation reflect the reality of how people are using composer? In my case, I'm using a hosting service where if I want to change a line of code on my production server, I have to build a new version of my containers. If that build includes a composer step, I've tied my ability to hotfix production to whether or not github is up. The easiest way to work around that is to commit my vendor directory. Yes, there are other ways around it, but they're clunky in comparison to simply _keeping all my code in the repo_. That is not an unreasonable choice. The documentation doesn't even acknowledge that it's possible, much less that it might be the right choice for some situations.

**Comment by Seldaek** [comment]
> Mine is bitter and pessimistic: software upgrades are a problem. You want to do them sparingly because any time you upgrade software it might introduce bugs and more regression testing is needed. Why change what's not broken?

The issue with this worldview IMO is that it will bite you in the ass in the long run. You end up doing bigger upgrades of more dependencies at once, so if something breaks you don't know where to start looking.

Doing frequent updates requires less jumping through hoops, and yes it carries a more frequent risk of something minor going wrong, but at least you'll be able to quickly revert that and analyze. With bigger updates it's not always possible to go back smoothly.

So I disagree this world view is beneficial, and I don't want to encourage anyone to go there.

> I use `composer require`, of course, in fact, I never use `composer update` itself exactly because of this, however that only works in the simplest of case: when there are no dependencies needed to be updated.

I hardly use `require` except to add new dependencies. If I want to update some, I run `composer update x y vendor/*`, eventually with `-w` or `-W` if needed but I rarely resort to that. That gives fine-grained control.

If you use `-W` then yes you are telling Composer to update all the things that package depends on, there is no way to just "fix dependencies if needed and stop", unfortunately that is a much more complex problem than unlocking dependencies and letting the solver find a workable solution. We could try to unlock one by one until it works, but that's not guaranteed to give you the solution you want either, and it'd require a multi-pass solving which adds a lot of complexity and CPU time for a case which I don't see as super needed considering running updated deps is generally a good thing. Plus if you absolutely need an older dep of something then you can require that explicitly (or conflict against new versions) to avoid updates.

----

@ksenzee I think https://getcomposer.org/doc/faqs/should-i-commit-the-dependencies-in-my-vendor-directory.md kinda acknolwedges that it's possible. I still don't think it's a good idea, and have never felt the need to do so, so I'd rather not encourage it any further, sorry.
